Azure/sap-hana
# Utility script support for multiple input JSON templates

## Problem

Changes introduced in #288 allow terraform to be easily run against a single JSON input template (single node HANA), but as more templates are added the scripts will need to allow for the user to make a choice on which to use (see #292).

The current scripts also write errors to STDOUT rather than STDERR (see #295).

## Description

This PR enables the V2 terraform script to be run using any input JSON template within the example templates directory, and provides the user with appropriate usage details and suggestions in the case when no template is given, or the given template does not exist. The USAGE guide is also updated to cover this interface change, and a new "minimal example" input JSON template is added `rti_only` that can be used to provision/destroy just the RTI instance. This can be used to test RTI-related provisioning, and runs in considerably less time than the SAP `single_node_hana` alternative.

The PR also fixes #295 by ensuring all error output created through the error output function is sent to STDERR.

**Note:** This PR does not cover the changes required to enable the `util/set_sap_download_credentials.sh` script to work with multiple input JSON templates. That feature will be added when more SAP input JSON templates are added.

## Pre-Requisites

1. Have followed the initial parts of [the USAGE guide up to section **Build/Update/Destroy Lifecycle**](https://github.com/Azure/sap-hana/blob/226cc37827333e5b96319f297c85313418f50144/deploy/v2/USAGE.md#buildupdatedestroy-lifecycle).
1. `shellcheck` installed on the workstation you're testing from (for test 3)

## Test Instructions

- [ ] 1. Review the [USAGE.md](https://github.com/Azure/sap-hana/pull/297/files#diff-7b2ac2a41f894bd0e6577ab6721d0d9d) guide updates to ensure they are clear.
- [ ] 2. Follow the usage guide to ensure the process works and the new utility scripts run successfully:
   - Potential alternative test cases:
      1. `util/terraform_v2.sh apply rti_only`
      1. `util/terraform_v2.sh destroy rti_only`
- [ ] 3. Shellcheck review the scripts by running `shellcheck util/*.sh` (no results should be shown and there should be a zero exit status from `echo $?` when run after shellcheck)
- [ ] 4. Check some edge cases for reasonable error handling in the `util/terraform_v2.sh` script, such as:
   - Not specifying command line arguments, when they're expected (e.g. `util/terraform_v2.sh`)
   - Specifying only partial command line arguments (e.g. `util/terraform_v2.sh apply`)
   - Specifying template files that don't exist  (e.g. `util/terraform_v2.sh apply test`)
- [ ] 5. Check that errors are now going to STDOUT by running through the following:
  - [ ] Check the following does not display the error "ERROR: 'test' is not a valid <JSON template name> to use with the 'destroy' option" on the terminal:
    ```bash
    util/terraform_v2.sh destroy test 2>error.log
    ```
  - [ ] Check the above error is in the `error.log` file:
    ```bash
    cat error.log
    ```
